context/llpcGraphicsContext.cpp

```diff
@@ -671,64 +667,186 @@ bool GraphicsContext::CheckGsOnChipValidity()
+// Does user data node merging for all shader stages
 void GraphicsContext::DoUserDataNodeMerge()
 {
-    const bool hasVs  = ((m_stageMask & ShaderStageToMask(ShaderStageVertex)) != 0);
-    const bool hasTcs = ((m_stageMask & ShaderStageToMask(ShaderStageTessControl)) != 0);
-    const bool hasTes = ((m_stageMask & ShaderStageToMask(ShaderStageTessEval)) != 0);
-    const bool hasGs  = ((m_stageMask & ShaderStageToMask(ShaderStageGeometry)) != 0);
+    uint32_t stageMask = GetShaderStageMask();
+    SmallVector<ResourceMappingNode, 8> allNodes;
+
+    // No need to merge if there is only one shader stage.
+    if (isPowerOf2_32(stageMask))
+    {
+        return;
+    }
+
+    // Collect user data nodes from all shader stages into one big table.
+    for (uint32_t stage = 0; stage < ShaderStageCount; ++stage)
+    {
+        if ((stageMask >> stage) & 1)
+        {
+            auto pShaderInfo = GetPipelineShaderInfo(ShaderStage(stage));
+            for (const ResourceMappingNode& node : ArrayRef<ResourceMappingNode>(pShaderInfo->pUserDataNodes,
+                                                                                 pShaderInfo->userDataNodeCount))
+            {
+                allNodes.push_back(node);
+            }
+        }
+    }
+
+    // Sort and merge.
+    ArrayRef<ResourceMappingNode> mergedNodes = MergeUserDataNodeTable(allNodes);
+
+    // Collect descriptor range values (immutable descriptors) from all shader stages into one big table.
+    SmallVector<DescriptorRangeValue, 8> allRanges;
+    for (uint32_t stage = 0; stage < ShaderStageCount; ++stage)
+    {
+        if ((stageMask >> stage) & 1)
+        {
+            auto pShaderInfo = GetPipelineShaderInfo(ShaderStage(stage));
+            for (const DescriptorRangeValue& range :
+                        ArrayRef<DescriptorRangeValue>(pShaderInfo->pDescriptorRangeValues,
+                                                       pShaderInfo->descriptorRangeValueCount))
+            {
+                allRanges.push_back(range);
+            }
+        }
+    }
+
+    // Sort them by set and binding, so we can spot duplicates.
+    std::sort(allRanges.begin(),
+              allRanges.end(),
+              [](const DescriptorRangeValue& left, const DescriptorRangeValue& right)
+              {
+                  if (left.set != right.set)
+                  {
+                      return left.set < right.set;
+                  }
+                  return left.binding < right.binding;
+              });
+
+    if (allRanges.empty() == false)
+    {
+        // Create a new table with merged duplicates.
+        m_allocDescriptorRangeValues = make_unique<SmallVector<DescriptorRangeValue, 8>>();
+        auto &mergedRanges = *m_allocDescriptorRangeValues;
+        ArrayRef<DescriptorRangeValue> ranges = allRanges;
+
+        while (ranges.empty() == false)
+        {
+            // Find the next block of duplicate ranges.
+            uint32_t numDuplicates = 1;
```

Comment:
numDuplicates -> duplicateCount

We often use "count" to replace "num" in LLPC as a convention.